File: sketch/ProgressBar.cpp
#include "Arduino.h"
#include "ProgressBar.h"

ProgressBar::ProgressBar(LiquidCrystal& lcd, uint8_t cols, uint8_t row) 
  : lcd(lcd), cols(cols), row(row) {
}

void ProgressBar::begin() {
  createCustomCharacters();
}

void ProgressBar::createCustomCharacters() {
  for (int i = 0; i < 6; i++) {
    uint8_t block[8];
    uint8_t pattern = createBlockPattern(i);
    for (int j = 0; j < 8; j++) {
      block[j] = pattern;
    }
    lcd.createChar(i, block);
  }
}

uint8_t ProgressBar::createBlockPattern(int filledBits) {
  return 0b11111 << (5 - filledBits);
}

void ProgressBar::displayProgress(double progressRate) {
  int totalBlocks = cols * 5;
  int progress = static_cast<int>(round(totalBlocks * progressRate));

  for (int i = 0; i < cols; i++) {
    int size = min(progress, 5);
    lcd.setCursor(i, row);
    lcd.write(static_cast<uint8_t>(size));
    progress -= size;
  }
}

File: lib/progressbar.py
from display import Display


def create_block_pattern(filled_bits: int):
    return 0b11111 << (5 - filled_bits)


class ProgressBar:
    def __init__(self, display: Display, columns: int, row: int):
        self.display = display
        self.columns = columns
        self.row = row

    def begin(self):
        self.__create_custom_characters()

    def display_progress(self, rate: float):
        self.display.move_cursor(0, self.row)
        total_blocks = self.columns * 5
        progress = int(round(rate * total_blocks))

        for i in range(self.columns):
            size = min(progress, 5)
            self.display.print_char(size)
            progress -= size

    def __create_custom_characters(self):
        for i in range(6):
            self.display.create_char(i, [create_block_pattern(i)] * 8)
